raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1002,6 +987,10 @@ private boolean handleFetchResponse(
             return handled.get();
         }
 
+        if (!quorum.hasLeader()) {
```

Comment:
This makes me wonder.. If the client becomes a candidate due to fetch timeout, and then it finally receives the fetch response with no error code, then previously we would crash on `quorum.followerStateOrThrow()` right? If yes it seems to be a bug. But I'm wondering whether this is the best fix: in this case, should that candidate really transit back to a follower, or should it just ignore the response?
2) Before this,

Reply:
Just to be clear, we have already validated the epoch from the fetch response in `maybeHandleCommonResponse`. If we had become a candidate, then we would bump the epoch and then we don't need to worry about fetch responses being received from the old epoch.

That said, I think this logic did leave a bit to be desired. I ended up removing the check in #402 and enhancing `maybeHandleCommonResponse` instead. If it's ok with you, I'll leave this as it is and we can discuss the improvements in #402.

Reply:
Cool, let's keep it as-is in this PR then.